internal/cache/file/downloader/job.go

```diff
@@ -226,196 +225,186 @@ func (job *Job) notifySubscribers() {
 	}
-
-	logger.Tracef("Job:%p (%s:/%s) downloaded till %v offset.", job, job.bucket.Name(), job.object.Name, job.status.Offset)
+	job.mu.Lock()
+	defer job.mu.Unlock()
 	err = job.fileInfoCache.UpdateWithoutChangingOrder(fileInfoKeyName, updatedFileInfo)
-	if err != nil {
-		err = fmt.Errorf("updateFileInfoCache: error while inserting into fileInfoCache %s: %w", updatedFileInfo.Key, err)
-		return
+	if err == nil {
+		job.status.Offset = downloadedOffset
+		// Notify subscribers if file cache is updated.
+		logger.Tracef("Job:%p (%s:/%s) downloaded till %v offset.", job, job.bucket.Name(), job.object.Name, job.status.Offset)
+		job.notifySubscribers()
+	} else {
+		err = fmt.Errorf("updateStatusOffset: error while updating offset: %v in fileInfoCache %s: %w", downloadedOffset, updatedFileInfo.Key, err)
 	}
 	return
 }
 
+// downloadObjectToFile downloads the backing object from GCS into the given
+// file and updates the file info cache. It uses gcs.Bucket's NewReader method
+// to download the object.
+func (job *Job) downloadObjectToFile(cacheFile *os.File) (err error) {
+	var newReader io.ReadCloser
+	var start, end, sequentialReadSize, newReaderLimit int64
+	end = int64(job.object.Size)
+	sequentialReadSize = int64(job.sequentialReadSizeMb) * cacheutil.MiB
+
+	// Each iteration of this for loop, reads ReadChunkSize size of range of the
+	// backing object from reader into the file handle and updates the file info
+	// cache. In case, reader is not present for reading, it creates a
+	// gcs.Bucket's NewReader with size min(sequentialReadSize, object.Size).
+	for start < end {
+		if newReader == nil {
+			newReaderLimit = min(start+sequentialReadSize, end)
+			newReader, err = job.bucket.NewReader(
+				job.cancelCtx,
+				&gcs.ReadObjectRequest{
+					Name:       job.object.Name,
+					Generation: job.object.Generation,
+					Range: &gcs.ByteRange{
+						Start: uint64(start),
+						Limit: uint64(newReaderLimit),
+					},
+					ReadCompressed: job.object.HasContentEncodingGzip(),
+				})
+			if err != nil {
+				err = fmt.Errorf("downloadObjectToFile: error in creating NewReader with start %d and limit %d: %w", start, newReaderLimit, err)
+				return err
+			}
+			monitor.CaptureGCSReadMetrics(job.cancelCtx, util.Sequential, newReaderLimit-start)
+		}
+
+		maxRead := min(ReadChunkSize, newReaderLimit-start)
+
+		// Copy the contents from NewReader to cache file.
+		offsetWriter := io.NewOffsetWriter(cacheFile, start)
+		_, err = io.CopyN(offsetWriter, newReader, maxRead)
+		if err != nil {
+			err = fmt.Errorf("downloadObjectToFile: error at the time of copying content to cache file %w", err)
+			return err
+		}
+
+		start += maxRead
+		if start == newReaderLimit {
+			err = newReader.Close()
+			if err != nil {
+				logger.Errorf("Job:%p (%s:/%s) error while closing reader: %v", job, job.bucket.Name(), job.object.Name, err)
+			}
+			newReader = nil
+		}
+
+		err = job.updateStatusOffset(start)
```

Comment:
Please add a comment why are we not returning from the above err.

Reply:
Added. PTAL